### ISSUE TITLE
Lint unnecessary disabled reporting comments

### ISSUE THREAD

**Issue by josh** [eslint/eslint#6190]
_Repost of #6188 as an Issue_

ESLint configuration comments are handy for temporarily allowing an exception or hack. But sometimes the hack is fixed but the comment is still left in place.

A clever way to implement this would be with a eslint rule itself. However this is error prone. The most obvious thing is that disabling all rules in a file would then disable this meta rule.

Similar to `allowInlineConfig`, this probably means it needs to be implemented outside of the rule processing code path. I took a stab at plumbing a similar CLI flag to enable these checks.

If this is something that won't be accepted, would you at least consider publicly exposing the data on `reportConfig`? Theres really no way to access this without monkey patching `api.report`. `api.getReportingConfig()` would probably be enough for my needs.

Related https://github.com/eslint/eslint/issues/5659 https://github.com/sindresorhus/xo/issues/103 https://github.com/bbatsov/rubocop/issues/1865

**Comment by kaicataldo** [comment]
Hey there - I believe this has already been discussed and it was decided that this doesn't belong in core, though it would be a great tool to add to the community. The issue also had a suggestion as to how this could be achieved! https://github.com/eslint/eslint/issues/5801

**Comment by josh** [comment]
> Just run eslint once with --no-inline-config and once without and compare the output.

Isn't good enough for me. That requires two passes of an entire codebase. Also, `--no-inline-config` disables `/* eslint env:0 */` configurable which should still be respected. I'm talking about `// eslint-disable` comments which behave much different. The linters on those lines still run and don't reported.

Its pretty silly to make another pass at the file when eslint already has all the information about which lines are skipped, but just won't expose it as a public api.

The only way I've figured out how to implement this is by monkey patching `api.verify` which is not an acceptable solution to me.

https://github.com/github/eslint-plugin-github/blob/master/lib/rules/no-unused-disabled-rules.js#L6

**Comment by nzakas** [comment]
@josh we can consider it, but there needs to be a lot more discussion. Your PR is pretty involved, and so we need to decide if that level of change is something we want to consider for this use case (which is pretty narrow).

To give some more structure to this conversation, here's what I'd like to ask you to do:

1. Clarify the problem you're trying to solve.
2. Explain the solution you have in mind
  - How would we know if a comment isn't necessary?
  - How would we tell people a comment isn't necessary? Is it a warning, an error, a fatal error?
  - Is a CLI option needed or should this be a default behavior?
3. Explain any alternatives to that solution
  - What is it you'd need to have exposed to create your own solution?

**Comment by josh** [comment]
> but there needs to be a lot more discussion.

Awesome! Thats all I hoped for. ❤️ 

> Your PR is pretty involved

Yeah, I don't like the implementation at all, but its a point of discussion.

> Clarify the problem you're trying to solve.

I want to ensure `// eslint-disable` comments are cleaned up from the codebase as issues are addressed. Its not always on the same line, so not obvious when you say fixed the last "no-var" case.

``` js
/* legacy file from coffescript conversion */
/* TODO: cleanup these eslint rules someday */
/* eslint-disable no-var */

// ...

function legacyCode() {
  var foo = 1
}

// ...

function newerFunctionAddedRecently() {
  const bar = 2
}
```

> How would we know if a comment isn't necessary?

At a high level, if a comment can be removed and no new errors would be reported.

Implementation wise, if a `reportConfig` is never used by `isDisabledByReportingConfig`.

> How would we tell people a comment isn't necessary? Is it a warning, an error, a fatal error?

This is the part where it being implemented as rule would be convenient at least. The `// eslint-disable` comment node itself could be reported. I don't think that exactly would be possible. But if it produced the same internal error `message` object, linting integration with Sublime or Atom wouldn't know the difference.

> Is a CLI option needed or should this be a default behavior?

I think the CLI option is ugly and would rather avoid it. I don't think it should be the default either. Configuring in the `.eslintrc.json` sounds ideally to me. But again, the implementation of this is rather different than regular rules. So wherever works me

> What is it you'd need to have exposed to create your own solution?

https://github.com/github/eslint-plugin-github/blob/master/lib/rules/no-unused-disabled-rules.js is the hack we're currently using. It needs to monkey patch some APIs which isn't ideal.

That approach could be implemented a little cleaner if some of the internals of `reportConfig` were exposed.

However, this still some deeper issues when its implemented as a rule itself. `whatever() // eslint-disable-line` without a `ruleId` will always prevent the comment itself from being reported since the entire line is excluded. Maybe that could be worked around by adding some `force: true` flag to `api.report`. In practice I don't care that much about that issue since we just don't use `// eslint-disable` without a specific rule. If it could be addressed, that'd be a nice to have.

Thanks for listening!

❤️ eslint

**Comment by nzakas** [comment]
Thanks for the details! I think the big question here is around user experience: do people want ESLint checking this all the time, or is it something people will use periodically?

And it seems like you're saying these problems should be reported as normal messages, but are they errors or warnings? Would we need a way to configure that?

**Comment by michaelficarra** [comment]
I want eslint checking this all the time. But I wouldn't mind if it was behind a configuration flag.

**Comment by josh** [comment]
> And it seems like you're saying these problems should be reported as normal messages, but are they errors or warnings? Would we need a way to configure that?

I personally want them as error, I want our CI to fail in this case. This again whats nice about it being a rule, you can configure it as an error or warning level.

Still thinking on a way it could be handled as a proper rule, I think we'd need a few new APIs. Some kind of `api.getRuleConfig`, some way to skipping `ruleConfig` via `api.report(... force: true)`. But this still the issue of when it runs on traversal, I'm using `Program:exit` right now, but surely some other rule could report after mine in an `Program:exit`. It all seems kinda error prone.

Seems like it just leads to it needing to be an eslint core check to ensure its really ran after all rules are processed.

Another idea for configuration:

``` json
{
  "rules": { ... }
  "no-unused-disabled-or-something": "warning",
  "no-unused-disabled-this-isnt-actually-a-normal-rule": "error"
}
```

**Comment by platinumazure** [comment]
I think this would definitely be cool to include in core and to have it be available behind a CLI option. However, I can see maintainability being a problem since `api.report()` is already so complex and all of the functions it calls are pretty sprawling.

I think this could be a prompt for a good refactoring of that space. Specifically, all the logic that checks if a particular node report location is within a disable section (as well as the data structures tracking where such disable sections are in the file) should be extracted to a new module, and then it would be easy to add code to `api.verify` which checks that all of the disable sections have at least one report on them.

**Comment by ilyavolodin** [comment]
I don't have a very strong opinion, but I don't think this should be in the core. I think it's a great use-case for additional tool, but linting our own comments is too meta for my taste. Now, I think we should expose functionality that would allow something like that to be done in a 3-rd party tool, but still not convinced at all about merging this into the core. This feels a lot like other requests that we've seen, like finding unconfigured rules, or auto-merging configs, and I still think it's a good use-case for 3-rd party tool.

**Comment by mysticatea** [comment]
> I want eslint checking this all the time. But I wouldn't mind if it was behind a configuration flag.

Me too.

- This is similar to `no-unused-vars` rule, I think this is a basic feature. 
- `no-unused-vars` rule has warning for `/*global */` comments. And `no-undef` rule has warning for `/*exported */` comments. This sounds a similar thing to me.
- Useless codes can be confusing us, so we have many rules to report useless codes, and this would be one of them. But a rule should not know other rules, so this is proper for core (or `onProblemReported` event?).
- We can make eslint missing problems by unintentional disable-comments. As we sometimes forget to remove `console.log`, we might forget to remove temporal disable-comments. Then it might cause problems in future. In my head, this feature is close to core than external tools.
- I remembered now, I wrote [a disable-comment](https://github.com/mysticatea/npm-run-all/blob/30c2949390b6fcf8cd7828ef054ce812877d3359/src/lib/index.js#L76-L77) before. This was stopping a false positive, but if I used this feature, the bug fix would break my CI builds. Wow, do I really want to use this feature always? Hmmm....

**Comment by nzakas** [comment]
I don't see a way that this functionality can be implemented as a rule without breaking some of the abstractions we have in place. For instance, rules intentionally cannot know about other rules or other rule configurations. That's important to keep in place to avoid improper coupling of rules. 

If anything, I think this would have to be a CLI flag. We can't start having people's builds failing because a disable comment isn't necessary.

**Comment by josh** [comment]
> If anything, I think this would have to be a CLI flag. We can't start having people's builds failing because a disable comment isn't necessary.

Would you 👍 reopening https://github.com/eslint/eslint/pull/6188, that was basically my approach there?

**Comment by nzakas** [comment]
@josh we should talk through some more of the details to see if we can get consensus on whether or not we want to do this before spending more time on code.

**Comment by platinumazure** [comment]
Wondering if anyone had any thoughts about [my comment above](https://github.com/eslint/eslint/issues/6190#issuecomment-219773700)? I think refactoring disabled rule logic will make it easier to determine how feasible this is in core and will hopefully improve maintainability.

**Comment by nzakas** [comment]
We have a lot of other, higher-priority work to do right now, so I wouldn't want to delve too deep into any major refactorings until we get further along on the JSCS compatibility work. 

I also think we are still at the phase of figuring out the correct user experience, so it doesn't make too much sense to dig into implementation details until after that.

**Comment by nzakas** [comment]
Per [TSC meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2016/07-Jul-2016.md#httpsgithubcomeslinteslintissues6190), we think there's value to this capability but would like to see a prototype before deciding whether or not to include it in the core. This would be to evaluate just how much code change is necessary and how it would integrate into the end-user experience.

Please use this issue to discussion possible end-user experiences and implementations.

**Comment by ilyavolodin** [comment]
Should original PR be considered as a prototype: #6188

**Comment by platinumazure** [comment]
That looks like a pretty good prototype. If it is rejected and I get a chance to refactor the disable report logic (see #6592), I'd be interested in taking a crack as well.

**Comment by btmills** [comment]
[I agree with](https://github.com/eslint/eslint/issues/6190#issuecomment-219764905) @michaelficarra in that I want this to run anytime I run ESLint. My ideal use case:

- Configure it like a rule (may lead to too many exceptions, but desirable if possible)
- Run with `eslint`
- Report like a rule, so it can even show up in editor integrations

**Comment by mysticatea** [comment]
My ideal usage (sorry, I could not respond in TSC meeting) are:

- I don't use it in CI build in order to not break the build by the patch release of ESLint.

  > I wrote [a disable-comment](https://github.com/mysticatea/npm-run-all/blob/30c2949390b6fcf8cd7828ef054ce812877d3359/src/lib/index.js#L76-L77) before. This was stopping a false positive, but if I used this feature, the bug fix would break my CI builds.
  > https://github.com/eslint/eslint/issues/6190#issuecomment-220007306

- I check it before commits in my local. (for all files.)
- I check it inside of my editor as same as ESLint rules. (for each file.)

I think it's fine to be worked in ESLint (core or plugin).

**Comment by nzakas** [comment]
Almost three months ago we said this needs a prototype to move forward. We still don't have one, so unless someone is willing to commit to doing one, I think we should close this.

**Comment by kaicataldo** [comment]
@platinumazure are you still considering working on this? Want to confirm before closing.

**Comment by platinumazure** [comment]
It's going to take me a while to get to, so let's close the issue and I'll
just write a PR when I get to this. Thanks for following up.

**Comment by mysticatea** [comment]
Just I found: https://github.com/github/eslint-plugin-github/blob/master/lib/rules/no-unused-disabled-rules.js

**Comment by mysticatea** [comment]
Today, I published [eslint-plugin-eslint-comments](https://github.com/mysticatea/eslint-plugin-eslint-comments) experimentally.

This plugin includes 6 rules to use ESLint directive-comments (e.g. `/*eslint-disable*/`) correctly. One of them is inspired by above GitHub's rule, but it has more correct behavior.

**Comment by nzakas** [comment]
What do we want to do here? Just say people should use a plugin? Or is someone interested in pursuing this in core?

**Comment by platinumazure** [comment]
@nzakas I like the idea of using the plugin for this particular use case.

I do still want to revisit how we handle report disabling (#6592), but this particular use case seems solved.

**Comment by kaicataldo** [comment]
Agreed. Given that this is solved by @mysticatea's plugin, I think it makes sense to point users to that.

**Comment by kaicataldo** [comment]
Closing this since @mysticatea's plugin solves the issue and there hasn't been any activity in quite a while.

For anyone looking for this behavior, please use [eslint-plugin-eslint-comments](https://github.com/mysticatea/eslint-plugin-eslint-comments)!
